esp-metadata/src/cfg/sha.rs

```diff
@@ -0,0 +1,164 @@
+use proc_macro2::Ident;
+use quote::{ToTokens, format_ident, quote};
+use serde::{Deserialize, Serialize};
+
+use crate::{generate_for_each_macro, number};
+
+struct ShaAlgo {
+    name: &'static str,
+    ident: Ident,
+    digest_words: u32,
+    block_size: u32,
+    message_len_bytes: u32,
+    insecure_against_collision: bool,
+    insecure_against_length_extension: bool,
+}
+
+#[derive(Default, Debug, Clone, Copy, Serialize, Deserialize)]
+pub struct ShaAlgoMap {
+    sha1: Option<u32>,
+    sha224: Option<u32>,
+    sha256: Option<u32>,
+    sha384: Option<u32>,
+    sha512: Option<u32>,
+    sha512_224: Option<u32>,
+    sha512_256: Option<u32>,
+    sha512_t: Option<u32>,
+}
+
+impl super::GenericProperty for ShaAlgoMap {
+    fn for_each_macro(&self) -> Option<proc_macro2::TokenStream> {
+        let modes = [
+            ("SHA-1", self.sha1),
+            ("SHA-224", self.sha224),
+            ("SHA-256", self.sha256),
+            ("SHA-384", self.sha384),
+            ("SHA-512", self.sha512),
+            ("SHA-512/224", self.sha512_224),
+            ("SHA-512/256", self.sha512_256),
+            ("SHA-512/t", self.sha512_t),
+        ];
+
+        let algos = modes
+            .into_iter()
+            .filter_map(|(name, mode)| mode.map(|m| (name, m)))
+            .filter_map(|(name, mode)| ShaAlgo::new(name).map(|name| (name, mode)))
+            .map(|(algo, mode)| {
+                let mode = number(mode);
+                quote! { #algo, #mode }
+            })
+            .collect::<Vec<_>>();
+
+        Some(generate_for_each_macro(
+            "sha_algorithm",
+            &[("algos", &algos)],
+        ))
+    }
+}
+
+impl ShaAlgo {
+    fn new(algo: &str) -> Option<Self> {
+        let known = [
+            Self {
+                name: "SHA-1",
+                ident: format_ident!("Sha1"),
+                digest_words: 20,
+                block_size: 64,
+                message_len_bytes: 8,
+                insecure_against_collision: true,
+                insecure_against_length_extension: true,
+            },
+            Self {
+                name: "SHA-224",
+                ident: format_ident!("Sha224"),
+                digest_words: 28,
+                block_size: 64,
+                message_len_bytes: 8,
+                insecure_against_collision: false,
+                insecure_against_length_extension: true,
```

Comment:
out of curiosity: where is the information from? I know about SHA-1 but is there a consolidated list of these?